### PR TITLE
[FLINK-16602][k8s] Rework the internal/external Service

### PR DESCRIPTION
**Pull request by zhengcanbin**
## What is the purpose of the change

At the moment we usually create two Services for a Flink application, one is the internal Service and the other is the so-called rest Service, the previous aims for forwarding request from the TMs to the JM, and the rest Service mainly serves as an external service for the Flink application. Here is a summary of the issues:

1. The functionality boundary of the two Services is not clear enough since the internal Service could also become the rest Service when its exposed type is ClusterIP.
2. For the high availability scenario, we create a useless internal Service which does not help forward the internal requests since the TMs directly communicate with the JM via the IP or hostname of the JM Pod.
3. Headless service is enough to help forward the internal requests from the TMs to the JM. Service of ClusterIP type would add corresponding rules into the iptables, too many rules in the iptables would lower the kube-proxy's efficiency in refreshing iptables while notified of change events, which could possibly cause severe stability problems in a Kubernetes cluster.

Therefore, we propose some improvements to the current design:

1. Clarify the functionality boundary for the two Services, the internal Service only serves the internal communication from TMs to JM, while the rest Service makes the Flink cluster accessible from outside. The internal Service only exposes the RPC and BLOB ports while the external one exposes the REST port.
2. Do not create the internal Service in the high availability case.
3. Use HEADLESS type for the internal Service.

## Verifying this change

This change added unit tests and can be verified on a real K8s cluster as follows:

1. Deploy a non-HA session cluster and check there are two dedicated Services created, the internal one has NONE CLUSTER-IP.
2. Deploy an HA session cluster and check only the rest Service is created.
3. Check that the internal Service only exposes the RPC and BLOB port while the external one exposes the REST port.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)

### CODE REVIEW

**Comment by zhengcanbin** [comment]
@TisonKun Travis gives a green light, kindly expect to see your reply.

**Comment by zhengcanbin** [comment]
@flinkbot run azure

**Comment by zhengcanbin** [comment]
Hi, @wangyang0918, would you like to take a look at your convinence?

**Comment by StephanEwen** [comment]
Big +1 to the concept of this change.

This unclear separation of internal/external service and the fact that the internal service is even created for HS setups (where it is not used) is an issue.

I have not checked the code of this PR, though.

**Comment by StephanEwen** [comment]
@flinkbot approve until consensus

**Comment by StephanEwen** [comment]
I am wondering if we also update the K8s templates [here](https://ci.apache.org/projects/flink/flink-docs-master/ops/deployment/kubernetes.html).

There is also no distinction between internal/external, HA or non-HA.
That the selector mixes internal and external can even be a security issue.

**Comment by zhengcanbin** [comment]
> I am wondering if we also update the K8s templates [here](https://ci.apache.org/projects/flink/flink-docs-master/ops/deployment/kubernetes.html).
> 
> There is also no distinction between internal/external, HA or non-HA.
> That the selector mixes internal and external can even be a security issue.

Thanks a lot for your attention, @StephanEwen.  You are right, we should also update the [K8s template](https://ci.apache.org/projects/flink/flink-docs-master/ops/deployment/kubernetes.html) for distinction and security. Since this PR is dedicated to the native K8s integration, I tend to create a new ticket for updating the template. WDYT?

**Comment by wangyang0918** [comment]
Moreover, i have tested this PR in a real K8s cluster with/without HA enabled. It works as expected. Great job.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/parameters/KubernetesJobManagerParameters.java [hunk 1]
```diff
@@ -112,4 +113,8 @@ public String getEntrypointClass() {
 	public KubernetesConfigOptions.ServiceExposedType getRestServiceExposedType() {
 		return flinkConfig.get(KubernetesConfigOptions.REST_SERVICE_EXPOSED_TYPE);
 	}
+
+	public boolean enableInternalService() {
```
[the remark this answers is not in the document]
Comment (zhengcanbin) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/parameters/KubernetesJobManagerParameters.java]: Agree!

**Comment by zhengcanbin** [comment]
Thanks for the comment @wangyang0918!

Could you help take another look at this PR? @StephanEwen @tillrohrmann @TisonKun

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/InternalServiceDecorator.java [hunk 2]
```diff
@@ -19,63 +19,67 @@
...
+		final String serviceName = KubernetesUtils.getInternalServiceName(kubernetesJobManagerParameters.getClusterId());
+
+		final Service headlessService = new ServiceBuilder()
+			.withApiVersion(Constants.API_VERSION)
+			.withNewMetadata()
+				.withName(serviceName)
+				.withLabels(kubernetesJobManagerParameters.getCommonLabels())
+				.endMetadata()
+			.withNewSpec()
+				.withClusterIP("None")
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/InternalServiceDecorator.java:65]: @zhengcanbin does this setting refer to `Headless service`? Can we move `"None"` into `Constants`?

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactoryTest.java [hunk 3]
```diff
@@ -179,8 +180,9 @@ public void testServices() {
 		final Service resultInternalService = internalServiceCandidates.get(0);
 		assertEquals(2, resultInternalService.getMetadata().getLabels().size());
 
-		assertEquals(resultInternalService.getSpec().getType(), "ClusterIP");
-		assertEquals(3, resultInternalService.getSpec().getPorts().size());
+		assertNull(resultInternalService.getSpec().getType());
+		assertEquals("None", resultInternalService.getSpec().getClusterIP());
```
Comment (TisonKun) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactoryTest.java:184]: Also update literal `"None"` here.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/InternalServiceDecorator.java (hunk 2 above)
Comment (zhengcanbin) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/InternalServiceDecorator.java]: > @zhengcanbin does this setting refer to `Headless service`?
Yes.

> Can we move `"None"` into `Constants`?
Fixed by [1188bfd](https://github.com/apache/flink/pull/11456/commits/1188bfdbd6ecde4c6277b93cea471636852ea9b9)

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactoryTest.java (hunk 3 above)
Comment (zhengcanbin) [reply flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactoryTest.java]: Fixed by [1188bfd](https://github.com/apache/flink/pull/11456/commits/1188bfdbd6ecde4c6277b93cea471636852ea9b9)

**Comment by zhengcanbin** [comment]
Thanks for the review @TisonKun! I have addressed your comments.

**Comment by zhengcanbin** [comment]
@flinkbot run travis

**Comment by zhengcanbin** [comment]
@flinkbot run azure

**Pull request merged**